lib/filterx/func-str.c

```diff
@@ -0,0 +1,459 @@
+#include "filterx/expr-literal.h"
+#include "filterx/expr-literal-generator.h"
+#include "filterx/filterx-eval.h"
+#include "filterx/object-extractor.h"
+#include "filterx/object-string.h"
+#include "filterx/object-list-interface.h"
+#include "scratch-buffers.h"
+
+#define FILTERX_FUNC_STARTSWITH_USAGE "Usage: startswith(string, prefix, ignorecase=true)" \
+"or startswith(string, [prefix_1, prefix_2, ..], ignorecase=true)"
+
+#define FILTERX_FUNC_ENDSWITH_USAGE "Usage: endswith(string, suffix, ignorecase=true)" \
+"or endswith(string, [suffix_1, suffix_2, ..], ignorecase=true)"
+
+#define FILTERX_FUNC_INCLUDES_USAGE "Usage: includes(string, substring, ignorecase=true)" \
+"or includes(string, [substring_1, substring_2, ..], ignorecase=true)"
+
+typedef struct _FilterXStringWithCache
+{
+  FilterXExpr *expr;
+  const gchar *str_value;
+  gsize str_len;
+} FilterXStringWithCache;
+
+typedef gboolean (*FilterXExprAffixProcessFunc)(const gchar *haystack, gsize haystack_len, const gchar *needle,
+                                                gsize needle_len);
+typedef struct _FilterXExprAffix
+{
+  FilterXFunction super;
+  gboolean ignore_case;
+  FilterXExpr *haystack;
+  struct
+  {
+    FilterXExpr *expr;
+    GPtrArray *cached_strings;
+  } needle;
+  FilterXExprAffixProcessFunc process;
+} FilterXExprAffix;
+
+static GString *
+_format_str_obj(FilterXObject *obj)
+{
+  GString *str = scratch_buffers_alloc();
+
+  if(!filterx_object_repr(obj, str))
+    return NULL;
+
+  return str;
+}
+
+static gboolean
+_do_casefold(GString *str)
+{
+  if(str->len > G_MAXSSIZE)
+    return FALSE;
+
+  gchar *casefolded_str = g_utf8_casefold(str->str, (gssize) str->len);
+  g_string_assign(str, casefolded_str);
+
+  g_free(casefolded_str);
+  return TRUE;
+}
+
+static GString *
+_expr_format(FilterXExpr *expr, gboolean ignore_case)
+{
+  FilterXObject *obj = filterx_expr_eval_typed(expr);
+  if (!obj)
+    return NULL;
+
+  GString *result = _format_str_obj(obj);
+  if (!result)
+    filterx_eval_push_error("failed to extract string value, repr() failed", expr, obj);
+  filterx_object_unref(obj);
+
+  if (ignore_case && !_do_casefold(result))
+    return NULL;
+
+  return result;
+}
+
+static GString *
+_obj_format(FilterXObject *obj, gboolean ignore_case)
+{
+  GString *result = _format_str_obj(obj);
+  if (!result)
+    filterx_eval_push_error("failed to extract string value, repr() failed", NULL, obj);
+  filterx_object_unref(obj);
+
+  if (ignore_case && !_do_casefold(result))
+    return NULL;
+
+  return result;
+}
+
+static gboolean
+_string_with_cache_fill_cache(FilterXStringWithCache *self, gboolean ignore_case)
+{
+  if(!filterx_expr_is_literal(self->expr))
+    return TRUE;
+
+  return !!_expr_format(self->expr, ignore_case);
+}
+
+static FilterXStringWithCache *
+_string_with_cache_new(FilterXExpr *expr, gboolean ignore_case)
+{
+  FilterXStringWithCache *self = g_new0(FilterXStringWithCache, 1);
+  self->expr = expr;
```

Comment:
Let's ref this here, and unref in the `free()` method (which is missing right now).